macro-based string interpolation formatter

dgruntz opened the pull request:
the following feature implements a macro-based string interpolator.
This way, the format strings are compared with the types of the
actual arguments. If they do not match, the compiler reports a
type error.

Review: @xeno-by @cvogt

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/424/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/424/

dgruntz commented:
The problem is that this change requires a new starr. I will change the commit so that it contains the old definition of StringContext.f and will ask @xeno-by on how to upload a new starr.

xeno-by commented:
You need credentials to run ./push-binary-libs.sh. Ask Josh about it. In the meanwhile, you can use mine. Check your email.

dgruntz commented:
Do I have to upload all, or only the compiler-jar/reflection-jar/library-jar?
Dominik

-----Original Message-----
From: Eugene Burmako [mailto:reply@reply.github.com] 
Sent: Donnerstag, 5. Juli 2012 17:18
To: Gruntz Dominik
Subject: Re: [scala] macro-based string interpolation formatter (#825)

You need credentials to run ./push-binary-libs.sh. Ask Josh about it. In the meanwhile, you can use mine. Check your email.

---

Reply to this email directly or view it on GitHub:
https://github.com/scala/scala/pull/825#issuecomment-6781469

xeno-by commented:
Upload all of them to be sure.

On 5 July 2012 19:24, dgruntz <
reply@reply.github.com

> wrote:
> 
> Do I have to upload all, or only the
> compiler-jar/reflection-jar/library-jar?
> Dominik
> 
> -----Original Message-----
> From: Eugene Burmako [mailto:reply@reply.github.com]
> Sent: Donnerstag, 5. Juli 2012 17:18
> To: Gruntz Dominik
> Subject: Re: [scala] macro-based string interpolation formatter (#825)
> 
> You need credentials to run ./push-binary-libs.sh. Ask Josh about it. In
> the meanwhile, you can use mine. Check your email.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/825#issuecomment-6781469
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/825#issuecomment-6781656

paulp commented on src/compiler/scala/tools/reflect/MacroImplementations.scala:
```diff
@@ -0,0 +1,146 @@
+
+  def macro_StringInterpolation_f(p: List[Tree], args: List[Tree]): Tree = {
+
+    val parts = p map { t => val Literal(Constant(s:String)) = t; s }
```
This will just crash on anything but a string literal; even if it isn't intended to be called by anything but, it should issue a reasonable error message if it happens.  It's definitely possible because it's a public method, and there's no way to write a type signature like "LiteralString => ..."

dgruntz replied:
Thanks for that comment. I just fixed that with pattern matching in 14f7cee.

dgruntz committed:
adds the sha1 files of the new starr / stringContext.f

This commit provides the new sha1 codes of the new STARR.
Moreover, it replaces the implementation of StringContext.f
to `macro ???`. The implementation is magically hardwired into
`scala.tools.reflect.MacroImplementations.macro_StringInterpolation_f`
by the new STARR.

xeno-by commented:
LVGTM

adriaanm commented:
please squash the commits in this pull request as they won't build invidually

adriaanm commented:
you can squash them using

```
git rebase -i d8c107c4a23543053a492ec53043c3b70a120940^ # edit the last two 'pick's in the rebase script to say 'squash'
git push origin -f # assuming origin is the remote for your github repo
```

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/439/

Commit:
macro-based string interpolation formatter

This commit provides a macro based string interpolation formatter.
The macro is implemented in MacroImplementations.scala.

In order to still be able to build a new STARR, the implementation
in StringContext.f is not yet changed. This will be replaced in
a later commit.

Commit:
stringinterpolation macro test files

This commit adds test files
neg: checks the error messages generated by the compiler
run: checks the macro implementation features

dgruntz commented:
PLS REBUILD ALL

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/439/

dgruntz committed:
removes special case for BigInteger

After rebasing, the commit 34d36108bf is in the history, therefore
a patch in my code can be removed as an implicit conversion from
BigInteger to BigInt was added in commit 34d36108bf.
The tests cases test this case and still pass.

dgruntz commented:
I cannot reproduce the test error which was reported by Jenkins. The problem is /files/jvm/actmig-loop-react.scala. When I execute that code locally, I get:

```
>test\partest test\files\jvm\actmig-loop-react.scala
Testing individual files
testing: [...]\files\jvm\actmig-loop-react.scala                      [  OK  ]
All of 1 tests were successful (elapsed time: 00:00:15)
```

When I compare the output which appears in the Jenkins log to the expected output, then the order of the lines differ (which can easily happen in a threaded test), but I haven't looked at the details.

How do I have to proceed in this situation?

xeno-by commented:
I guess this is a spurious error. @lrytz, @vjovanov, could you, please,
take a look?

On 8 July 2012 00:42, dgruntz <
reply@reply.github.com

> wrote:
> 
> I cannot reproduce the test error which was reported by Jenkins. The
> problem is /files/jvm/actmig-loop-react.scala. When I execute that code
> locally, I get:
> 
> ```
> >test\partest test\files\jvm\actmig-loop-react.scala
> Testing individual files
> testing: [...]\files\jvm\actmig-loop-react.scala
> ```
> 
>  [  OK  ]
>     All of 1 tests were successful (elapsed time: 00:00:15)
> 
> When I compare the output which appears in the Jenkins log to the expected
> output, then the order of the lines differ (which can easily happen in a
> threaded test), but I haven't looked at the details.
> 
> How do I have to proceed in this situation?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/825#issuecomment-6826383

dgruntz commented:
I looked at the test case and came to the conclusion that it is not deterministic. The problem is in test method `testNestedReact`:

```
  loopWhile(c) {
    react {
      case x: Int =>
        // do task
        println("do task " + x)
        if (x == 42) {
          c = false
          finishedTNR1.success(true)
        } else
          react {
            case y: String =>
              println("do string " + y)
          }
        println("after react")
    }
  }
```

When the actor receives 42, it calls success on the promise `finishedTNR1` which will trigger the second test, but simultaneously, this thread will also print `after react`. Thus, this print statement is executed in parallel to the second part of this test.

This problem could be fixed by moving `finishedTNR1.success(true)` beyond `println("after react")`.

adriaanm commented:
PLS REBUILD ALL

(Vojin removed the actor test)

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/467/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/467/

retronym commented:
@dgruntz I assume you're working on a macro version of `"regexes (rock|suck)".r` now, too? :)

dgruntz commented:
will consider that, but first I go for two weeks in holidays. I also plan to improve the f-macro and to check the format strings (e.g. `f"${4}%3.f"` gives a runtime-error -- could be checked in the macro as well), but then I would like to have the positions of the part strings, this is not yet tracked by the compiler.
